clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java

```diff
@@ -202,59 +206,56 @@ public String toString() {
          * Mark that we want data from this partition in the upcoming fetch.
          */
         public void add(TopicPartition topicPartition, PartitionData data) {
+            if (sessionPartitions.put(topicPartition, data) != null) {
```

Comment:
If we weren't tracking altered and added for debugging purposes, I would suggest we only call `next.put` here and then perform a `sessionPartitions.putAll(next)` on build. putAll is a bit more efficient, but I'm not sure it matters enough to change this given that we wouldn't be able to track altered and added. A second question is whether debug logging is a good enough reason to maintain altered and added for debugging alone.